lib/tree.rb

```diff
@@ -2,54 +2,130 @@ class TreeNode
   attr_reader :key, :value
   attr_accessor :left, :right
 
-   def initialize(key, val)
+  def initialize(key, val)
     @key = key
     @value = val
     @left = nil
     @right = nil
-   end
+  end
 end
 
 class Tree
   attr_reader :root
+  
   def initialize
     @root = nil
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
-  def add(key, value)
-    raise NotImplementedError
+  def add_helper(current_node, new_node)
+    return new_node if current_node.nil?
+
+    if new_node.key < current_node.key
+      current_node.left = add_helper(current_node.left, new_node)
+    elsif new_node.key > current_node.key
+      current_node.right = add_helper(current_node.right, new_node)
+    end 
+
+    return current_node
+  end 
+
+  # Time Complexity: o(log(n))
+  # Space Complexity: o(1)
+  def add(key, value = nil)
```

Comment:
👍 the time complexity is O(log n) if the tree is balanced and O(n) if it's unbalanced.  The space complexity is the **same** due to recursion.